Issue361 units validation

kerimoyle opened the pull request:
Fourth in the series of smaller pull requests replacing #304 and addressing issue #361.

Unit validation:

- [x] checks for cyclical dependencies of units
- [x] checks that units are the same between equivalent variables
- [x] tests
- [x] coverage

kerimoyle committed:
Add validation for cyclic units

- cyclic unit dependencies are checked
- loops are alerted
- any depth of inheritance is ok
- multipliers are not validated, but the code to do it is present and commented out

kerimoyle committed:
Update validator.cpp

kerimoyle committed:
Fixed tests

kerimoyle committed:
Merge branch 'develop' into issue295_units_validation

kerimoyle committed:
Fixing merge conflicts

kerimoyle committed:
Merge pull request #15 from cellml/develop

Updating from libcellml/develop

kerimoyle committed:
Merge branch 'develop' into issue361_units_validation

kerimoyle committed:
Validator tests added for coverage

kerimoyle committed:
Added coverage tests for out of range integer prefix

kerimoyle commented:
@nickerso @hsorby @agarny Ready for review please.

kerimoyle committed:
Merge remote-tracking branch 'upstream/develop' into develop

kerimoyle committed:
Merge branch 'develop' into issue361_units_validation

hsorby reviewed (dismissed):
I find the tests too long and that they lack focus.  I would like to see tests that start simple and get more complicated.

* First test should be does this variables units match this other variables units.
* We only need to test one cycle if the cycles are not fundamentally different.
* What happens when you test the same variable to itself?

agarny commented on .gitignore:
```diff
@@ -16,5 +16,9 @@ CMakeLists.txt.user
+
+# VS Code configuration files
 .vscode/
+
```
FWIW, this file shouldn't be part of this PR. It has already been modified as a result of another PR.

kerimoyle committed:
Split validator tests into smaller pieces

kerimoyle commented:
> I find the tests too long and that they lack focus. I would like to see tests that start simple and get more complicated.

Is there one in particular you mean?  The cycle network is as simple as they can be to actually test the functionality of the validator.  I've split that test in half now so that one is valid and the other not.

> - First test should be does this variables units match this other variables units.

I've moved `TEST(Validator, variableEquivalentUnits)` which does exactly that to be the first of the bunch. It has a lot of different cases - starting super-simple and getting more interesting - but I've split them out into separate tests which hopefully makes it easier to follow ...

> - We only need to test one cycle if the cycles are not fundamentally different.

Do you mean that units cycles are reported more than once?  There's only one cycle there but I'll see if I can suppress reporting of the same one more than once - thanks for the heads-up!

> - What happens when you test the same variable to itself?

That's a variable validation thing rather than units.  I'll add it in to #350 where the variable network is tested.

I've also added the individual standard units testing that @agarny wanted from #340 there too.

kerimoyle committed:
Added individual tests for base unit to standard unit equivalence

Addresses #340

kerimoyle commented:
Coverage test is failing at 100% again?

kerimoyle committed:
Merge pull request #16 from cellml/develop

Upgrading coverage testing

kerimoyle committed:
Merge branch 'develop' into issue361_units_validation

kerimoyle committed:
Removed US spellings: deka, liter, meter

Addresses issue #327 and https://github.com/cellml/cellml-specification/issues/21

kerimoyle commented:
Removed American spellings of "deka", "liter" and "meter" from the accepted units and prefixes.  They will return a validation error now.  Addresses #327 and https://github.com/cellml/cellml-specification/issues/21

kerimoyle committed:
Update Ohara_Rudy_2011.cellml

kerimoyle committed:
One line function call for coverage

hsorby committed:
Update units.h

Remove comment on alternate spelling.

kerimoyle committed:
Single reporting for each unit loop found in validator

hsorby reviewed (commented):
I don't think we need to go around changing the units in the Ohara Rudy model where it isn't necessary.  Names of units can have whatever spelling is desired we only need the base units to be correctly spelt.

kerimoyle commented:
I just figured it was easier and more consistent to do it here, esp if it's used as an example.  I can change it back.

kerimoyle committed:
Merge pull request #17 from cellml/develop

Update from upstream

kerimoyle committed:
Merge branch 'develop' into issue361_units_validation

kerimoyle committed:
Update Ohara_Rudy_2011.cellml

Changed OharaRudy model back to having metRE, micrometER, centimetER, etc - only base units with correct spelling.

kerimoyle commented:
@hsorby Changed back to be correct spellings for only the base unit of metre: others are now centimeter, micrometer etc. @agarny @nickerso Good to go or more needed?

kerimoyle committed:
Merge branch 'develop' into issue295_units_validation

kerimoyle commented:
Nudge ... @hsorby @nickerso

hsorby commented:
Looking at this starting now, sorry for taking so long to get here.

hsorby reviewed (commented):
I still find the build up of tests missing.  I think the first test should be looking at testing the simplest possibly case, I don't find that with the tests written here.  In [this PR](https://github.com/kerimoyle/libcellml-keri/pull/20) I have removed a chunk of code that had absolutely zero effect on the test in question.  Unit tests are hard enough to read without adding unnecessary statements.

Apart from the above, I think I am likely to approve these changes if the PR above is merged in.

kerimoyle commented:
Thanks for the review @hsorby .  I agree that the simplest test case was missing, but I do think that it was important to test for cycles in the presence of branches - which was the point of the complicated (zero effect) bit of the last test.  It's more for testing that the valid branches to which the cycle is connected are not included in the reporting.  If you'd rather, I can move this to another test case, leaving both your simpler one and my branched one separate?

hsorby commented:
Yes the simple test should be there but if you feel you need a test to show that it works with branches the certainly do so in addition to that one.

kerimoyle committed:
Requested test changes made to validator unit cycles

kerimoyle commented:
All done :) Can be merged now @hsorby @nickerso @agarny ?

hsorby reviewed (dismissed):
I have had a look and a play with this and find that it is doing the job as advertised. So I am approving this PR.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

hsorby commented on tests/resources/Ohara_Rudy_2011.cellml:
```diff
@@ -61,28 +61,28 @@ http://www.cellml.org/
         <unit prefix="micro" units="farad"/>
     </units>
-    <units name="micrometer_3">
+    <units name="micrometre_3">
```
We didn't need to change these (and others like it) as they are user defined.  The user is free to use whichever spelling they prefer.

agarny replied:
Agreed (I saw your previous comment about this above), it's just that the name of some units had already been modified (to have "litre" rather than "liter"), so I thought I would just do the same for metre. However, since I agree with your point, I am going to revert the name of all those variables. We indeed only need the value of `units` to use the correct spelling.

agarny committed:
Some minor cleaning up.

agarny committed:
Various cleaning ups.

- Fixed some typos in the description of some methods.
- Added some full stops in the description of some methods.
- Added a description (to be checked!) for `multMap`.
- Make sure that we respect our naming convention, e.g. `unitMap` rather than `unitmap`.
- Removed some commented out code and some TODOs.
- Removed the declaration of some unused variables.

hsorby commented on src/validator.cpp:
```diff
@@ -29,9 +29,7 @@ limitations under the License.
 
 #include <cmath>
-
 #include <iostream>
```
This include may not be even needed?

agarny replied:
Good point, I haven't checked, but I am going to in a minute.

agarny committed:
Some minor cleaning up.

agarny committed:
Cleaning up: we don't need to compute and pass `multMap` around.

agarny committed:
Validator: removed unneeded header files.

agarny committed:
Make Windows happy.

agarny committed:
Make Linux happy.

agarny committed:
Validator: merged `incrementBaseUnitCount()` and `decrementBaseUnitCount()` into one method (`updateBaseUnitCount()`).

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Validator: various minor improvements to the code.

Only declare variables when actually necessary.

hsorby commented on src/validator.cpp:
```diff
@@ -1183,145 +1156,89 @@ bool Validator::ValidatorImpl::unitsAreEquivalent(const ModelPtr &model,
+            std::string myId;
+            double myExp;
+            double myMult;
+            double m;
```
Not super keen on this varabile name

kerimoyle replied:
The m? or the "my" ones?

kerimoyle replied:
Done (I think) :)

hsorby reviewed (dismissed):
Apart for very minor varialbe name this is still good for me.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -1065,3 +1065,726 @@ TEST(Validator, validMathCnElements)
+    const std::vector<std::string> expectedErrors = {
+        "Units reference 'meter' in units 'testunit2' is not a valid reference to a local units or a standard unit type.",
+        "Variable 'tomayto' has units of 'testunit1' and an equivalent variable 'tomahto' with non-matching units of 'testunit2'. The mismatch is: metre^1, ",
+        "Variable 'tomahto' has units of 'testunit2' and an equivalent variable 'tomayto' with non-matching units of 'testunit1'. The mismatch is: metre^-1, "};
```
The last two error messages are effectively the same. So, only one of them ought to be sufficient.

Also, the last two error messages end up with `, `, which is clearly not what we want. We want a properly formatted error message.

kerimoyle replied:
Done :)

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny commented on src/validator.cpp:
```diff
@@ -231,6 +225,54 @@ struct Validator::ValidatorImpl
+    * have equivalent units.
+    * @cellml2_19 Validate that equivalent variable pairs in the @p model
+    * have equivalent units.
+    * Any errors will be logged in the @c Validator.
```
What is that `@cellml2_19` about?

kerimoyle replied:
Ah, well spotted! That's a leftover from where I was coding all the validator tests to make sure that everything in the spec was covered.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -1065,3 +1063,726 @@ TEST(Validator, validMathCnElements)
+    // m^6
+    libcellml::UnitsPtr u4 = std::make_shared<libcellml::Units>();
+    u4->setName("u4");
+    u4->addUnit("u3", 15, 1.0); // standard, prefix, exponent.
```
I am confused here... shouldn't it be `18` rather than `15`? Or is the idea just to have variables with equivalent units as opposed to exactly the same unit?

agarny replied:
In fact, the current validation code seems to be checking only for equivalent units (and rightly so when it comes to validation), which means that we could in fact use any prefix we want here.

kerimoyle replied:
Hi @agarny Yes, apparently we don't mind about the multiplier/prefix mismatch of units across equivalent variables.  It will be raised as a "warning" level when these are in and I can finally get to working on the Issues structure for non-fatal stuff.

agarny replied:
Fair enough, however it might be worth making a small comment about this in the test. It's just that when I saw it, I got confused and couldn't help myself but working out the units equivalence by hand to make sure that it was "correct" (hence wasting a bit of time).

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

Also merged together the `unitNoCyclesBranching` and `unitCyclesFound` tests since they were effectively the same, apart from a couple of more lines for the latter test. So, now, we have a "new" test called `unitComplexCycle`.

agarny commented on tests/validator/validator.cpp:
```diff
@@ -1065,3 +1063,726 @@ TEST(Validator, validMathCnElements)
+        "Variable 'v3' has units of 'litre' and an equivalent variable 'v4' with non-matching units of 'gram'. The mismatch is: kilogram^-1, metre^3, ",
+        // "Variable 'v3' has units of 'litre' and an equivalent variable 'v9' with non-matching units of 'big_barrel'. The mismatch is: multiplication factor of 10^-3, ", // NOTE: removed until multiplication checking is turned on.
+        "Variable 'v6' has units of 'second' and an equivalent variable 'v5' with non-matching units of 'metre'. The mismatch is: metre^-1, second^1, ",
+        // "Variable 'v9' has units of 'big_barrel' and an equivalent variable 'v3' with non-matching units of 'litre'. The mismatch is: multiplication factor of 10^3, ", // NOTE: removed until multiplication checking is turned on.
```
Why are there still a couple of expected errors that are commented out? Am I to understand that this PR is not complete in fact? If so, then I guess that commit 14ea753 should be reverted?...

Also, like in another test, the expected errors are effectively duplicated, which is more confusing than anything. I would really have only one expected error between say `v1` and `v2` rather than one between `v1` and `v2` and another between `v2` and `v1`.

kerimoyle replied:
It's just as the comments say ... removed until multiplication checking is turned on.

agarny replied:
Sure, but... why isn't multiplication checking part of this PR?

agarny replied:
Ok, I imagine this is because we can only generate errors at this stage, as opposed to issues (which may be of either error, warning or hint type). Fair enough, but this means that this piece of commented out code should be removed. Then, if there isn't already one, a GitHub issue should be created to explain the additional work that will need to be done once issues are available.

agarny reviewed (changes requested):
This looks good overall, pending the few issues I have raised.

kerimoyle reviewed (commented):
I'm confused!  I thought that tests had to be as small as possible, even where they were duplicating much the same stuff.  This has been bigger earlier and was then made smaller ... it would be nice to have consensus on what I was aiming for ... ?

kerimoyle committed:
Requested changes made

- variable naming
- removed @cellml2 reference
- removed duplicated reporting
- removed trailing comma from end of statement

kerimoyle commented:
Changes made as requested, should be good to go now? @agarny @hsorby @nickerso

nickerso commented:
> I'm confused! I thought that tests had to be as small as possible, even where they were duplicating much the same stuff. This has been bigger earlier and was then made smaller ... it would be nice to have consensus on what I was aiming for ... ?

I agree. I think that as the capabilities of libCellML have grown we have not be particularly consistent in the way that we write tests. I'd be inclined to not worry too much at present and once we catch up this kind of discussion can happen as part of a general code review.

agarny committed:
Some minor cleaning up.

agarny reviewed (commented):
I am not clear why multiplication checking is not part of this PR, but I am going to assume that it's just missing something here. So, apart from that, I am now fine with this PR.

hsorby commented on src/validator.cpp:
```diff
@@ -1183,145 +1156,89 @@ bool Validator::ValidatorImpl::unitsAreEquivalent(const ModelPtr &model,
+            std::string myId;
+            double myExp;
+            double myMult;
+            double m;
```
Yes, I don't like the `m` variable it seems it also means the `myMult` is also a bit disingenuous.

agarny replied:
All fine now.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny commented:
@hsorby and @nickerso, you might want to quickly review this PR and, if you are still happy with it, merge it.

hsorby reviewed (approved):
Looks good to me.

kerimoyle reviewed (commented):
Just a note that we *do* need to re-include the multMap to enable the multiplication checking later ...

agarny commented:
Indeed and that should be done as part of issue #381.
